lib/WebService/Async/CustomerIO.pm

```diff
@@ -0,0 +1,313 @@
+            limit    => RPS_LIMIT_TRACKING,
+            interval => 1,
+        );
+        $self->add_child($rl);
+
+        $rl;
+    };
+    $self->{api_ratelimiter} ||= do {
+        my $rl = WebService::Async::CustomerIO::RateLimiter->new(
+            limit    => RPS_LIMIT_API,
+            interval => 1,
+        );
+        $self->add_child($rl);
+
+        $rl;
+    };
+
+    $self->next::method(%args);
+}
+
+=head2 site_id
+
+=cut
+
+sub site_id {shift->{site_id}}
+
+=head2 api_key
+
+=cut
+
+sub api_key {shift->{api_key}}
+
+=head2 API endpoints:
+
+There is 2 stable API for Customer.io, if you need to add a new method check 
+the L<documentation for API|https://customer.io/docs/api/> which endpoint 
+you need to use:
+
+=over 4
+
+=item * C<Tracking API> - Behavioral Tracking API is used to identify and track
+customer data with Customer.io.
+
+=item * C<Regular API> - Currently, this endpoint is only used for sending
+L<API triggered broadcasts|https://customer.io/docs/api-triggered-broadcast-setup>.
+
+=back
+
+=head2 tracking_request
+
+Sending request to Tracking API end point.
+
+Usage: C<< tracking_request($method, $uri, $data) -> future($data) >>
+
+=cut
+
+sub tracking_request {
+    my ($self, $method, $uri, $data) = @_;
+    return $self->tracking_ratelimiter->acquire->then(sub {
+        $self->_request($method, join(q{/} => (TRACKING_END_POINT, $uri)), $data);
+    });
+}
+
+=head2 api_request
+
+Sending request to Regular API end point.
+
+Usage: C<< api_request($method, $uri, $data) -> future($data) >>
+
+=cut
+
+sub api_request {
+    my ($self, $method, $uri, $data) = @_;
+
+    return $self->api_ratelimiter->acquire->then(sub {
+        $self->_request($method, join(q{/} => (API_END_POINT, $uri)), $data);
+    });
+}
+
+=head2 api_ratelimiter
+
+Getter returns RateLimmiter for regular API endpoint.
+
+=cut
+
+sub api_ratelimiter {shift->{api_ratelimiter}}
+
+=head2 tracking_ratelimiter
+
+Getter returns RateLimmiter for tracking API endpoint.
+
+=cut
+
+sub tracking_ratelimiter {shift->{tracking_ratelimiter}}
+
+sub _request {
+    my ($self, $method, $uri, $data) = @_;
+
+    my $body = $data             ? encode_json_utf8($data)
+             : $method eq 'POST' ? q{}
+             :                     undef;
+
+    return $self->_ua->do_request(
+            method  => $method,
+            uri     => $uri,
+            user    => $self->site_id,
+            pass    => $self->api_key,
+            !defined $body ? () : (
+                content      => $body,
+                content_type => 'application/json',
+            ),
+    )->catch(sub {
+        my ($code_msg, $err_type, $response) = @_;
+        return Future->fail(@_) unless $err_type && $err_type eq 'http';
+
+        my $code = $response->code;
+        my $request_data = {method => $method, uri => $uri, data => $data};
+
+        return Future->fail('RESOURCE_NOT_FOUND', 'customerio', $request_data)  if $code == 404;
```

Comment:
Would it make sense to create mapping of `error_code => error` like
```
my $error_mapping = { 404 => 'RESOURCE_NOT_FOUND', 400 => 'INVALID_REQUEST', ...};

$error_mapping->{$code} ?  return Future->fail($error_mapping->{$code}, 'customerio', $request_data) : return Future->fail('UNEXPECTED_HTTP_CODE: ' . $code_msg, 'customerio', $response)
```

Reply:
Good idea, thanks. Fixed.